cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -52,57 +54,283 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
+	if count > 1 {
 		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
 	}
 	return allErrs
 }
 
-func ValidateFileDiscovery(c *kubeadm.FileDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateArgSelection(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateTLSBootstrapTokenArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryFileArg(cfg, field.NewPath("discovery"))...)
+	allErrs = append(allErrs, ValidateDiscoveryURLArg(cfg, field.NewPath("discovery"))...)
+	if len(cfg.Token) != 0 {
+		cfg.TLSBootstrapToken = cfg.Token
+		cfg.DiscoveryToken = cfg.Token
+	}
+	return allErrs
+}
+
+func ValidateTokenArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.Token) != 0 {
+		if len(cfg.TLSBootstrapToken) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryURL) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--token is mutually exclusive with --tls-bootstrap-token, --discovery-token, --discovery-file, and --discovery-url",
+				),
+			)
+		}
+		if len(cfg.Masters) < 1 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--token must also specificy --masters",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func ValidateDiscoveryTokenArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryToken) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryURL) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-token is mutually exclusive with --token, --discovery-file, and --discovery-url",
+				),
+			)
+		}
+		if len(cfg.Masters) < 1 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-token must also specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-token must also specificy --tls-bootstrap-token",
+				),
+			)
+		}
+	}
+	return allErrs
+}
+
+func ValidateTLSBootstrapTokenArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.TLSBootstrapToken) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryURL) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--tls-bootstrap-token is mutually exclusive with --token, --discovery-file, and --discovery-url",
```

Comment:
nit on every error here; can you use TLSBootstrapToken like the field name instead of the flag?